src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,279 @@
+    input_files=merged_input_files,
+    output_files=('.coverage',),
+    description=f'Merge coverage reports.',
+  )
+
+  result = await Get[ExecuteProcessResult](
+    ExecuteProcessRequest,
+    request
+  )
+  return MergedCoverageData(coverage_data=result.output_directory_digest)
+
+
+COVERAGE_PLUGIN_MODULE_NAME = '__coverage_coverage_plugin__'
+
+
+def get_coverage_plugin_input():
+  return InputFilesContent(
+    FilesContent(
+      (
+        FileContent(
+          path=f'{COVERAGE_PLUGIN_MODULE_NAME}.py',
+          content=pkg_resources.resource_string(__name__, 'coverage/plugin.py'),
+          is_executable=False,
+        ),
+      )
+    )
+  )
+
+
+def ensure_section(config_parser: configparser.ConfigParser, section: str) -> None:
+  """Ensure a section exists in a ConfigParser."""
+  if not config_parser.has_section(section):
+    config_parser.add_section(section)
+
+
+def construct_coverage_config(source_roots, python_files) -> str:
+  # A map from source root stripped source to its source root. eg:
+  #  {'pants/testutil/subsystem/util.py': 'src/python'}
+  # This is so coverage reports referencing /chroot/path/pants/testutil/subsystem/util.py can be mapped
+  # back to the actual sources they reference when merging coverage reports.
+  source_to_target_base: Dict[str, str] = {}
+  for file_name in python_files:
+    source_root = source_roots.find_by_path(file_name)
+    source_root_stripped_path = file_name[len(source_root.path) + 1:]
+    source_to_target_base[source_root_stripped_path] = source_root.path
+
+  config_parser = configparser.ConfigParser()
+  config_parser.read_file(StringIO(DEFAULT_COVERAGE_CONFIG))
+  ensure_section(config_parser, 'run')
+  config_parser.set('run', 'plugins', COVERAGE_PLUGIN_MODULE_NAME)
+  config_parser.add_section(COVERAGE_PLUGIN_MODULE_NAME)
+  config_parser.set(COVERAGE_PLUGIN_MODULE_NAME, 'source_to_target_base', json.dumps(source_to_target_base))
+  config = StringIO()
+  config_parser.write(config)
+  return config.getvalue()
+
+
+def get_file_names(all_target_adaptors):
+  def iter_files():
+    for adaptor in all_target_adaptors:
+      if hasattr(adaptor, 'sources'):
+        for file in adaptor.sources.snapshot.files:
+          if file.endswith('.py'):
+            yield file
+
+  return list(iter_files())
+
+
+
+@rule(name="Generate coverage report")
+async def generate_coverage_report(
+  test_results: AddressAndTestResults,
+  transitive_targets: TransitiveHydratedTargets,
+  python_setup: PythonSetup,
+  coverage_setup: CoverageSetup,
+  coverage_toolbase: CoverageToolBase,
+  source_root_config: SourceRootConfig,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment,
+) -> PytestCoverageReport:
+  """Takes all python test results and generates a single coverage report in dist/coverage."""
+  requirements_pex = coverage_setup.requirements_pex
+  merged_coverage_data = await Get[MergedCoverageData](AddressAndTestResults, test_results)
```

Comment:
I believe you request `MergedCoverageData` directly in the rule signature, rather than awaiting it. The engine will then figure out to first run the merge rule and inject this type.